File: protocol/frame.go
// Package protocol implements the GS Protocol (GSP1) frame encoding and decoding.
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

const (
	// Magic is the protocol magic number "GSP1" (0x47535031).
	Magic uint32 = 0x47535031
	// Version is the current protocol version.
	Version uint8 = 1

	// HeaderSize is the fixed size of a frame header in bytes.
	HeaderSize = 12
)

// Frame type constants.
const (
	TypeConnect uint8 = 1
	TypeData    uint8 = 2
	TypePing    uint8 = 3
	TypePong    uint8 = 4
	TypeClose   uint8 = 5
	TypeError   uint8 = 6
	TypeHTTPReq  uint8 = 9
	TypeHTTPResp uint8 = 10
)

// Error codes returned in ERROR frames.
const (
	ErrAuthFailed      uint16 = 1001
	ErrInvalidTarget   uint16 = 1002
	ErrConnectFailed   uint16 = 1003
	ErrRateLimited     uint16 = 1004
	ErrInvalidFrame    uint16 = 1005
	ErrInternal        uint16 = 1006
)

// Frame represents a GSP1 protocol frame.
type Frame struct {
	Version  uint8
	Type     uint8
	Flags    uint16
	StreamID uint32 // v2 only
	Payload  []byte
}

// ConnectPayload holds the CONNECT frame payload fields.
type ConnectPayload struct {
	Host     string
	Port     uint16
	Password string
}

// ErrorPayload holds the ERROR frame payload fields.
type ErrorPayload struct {
	Code    uint16
	Message string
}

var (
	// ErrInvalidMagic is returned when the magic number does not match.
	ErrInvalidMagic = errors.New("protocol: invalid magic")
	// ErrInvalidVersion is returned when the protocol version is unsupported.
	ErrInvalidVersion = errors.New("protocol: unsupported version")
	// ErrPayloadTooLarge is returned when payload exceeds the maximum size.
	ErrPayloadTooLarge = errors.New("protocol: payload too large")
)

// MaxPayloadSize is the maximum allowed payload size (16 MB).
const MaxPayloadSize = 16 * 1024 * 1024

// EncodeConnectPayload serializes a CONNECT payload.
func EncodeConnectPayload(p ConnectPayload) []byte {
	hostBytes := []byte(p.Host)
	passBytes := []byte(p.Password)
	buf := make([]byte, 2+len(hostBytes)+2+2+len(passBytes))
	off := 0
	off += putString(buf[off:], hostBytes)
	binary.BigEndian.PutUint16(buf[off:], p.Port)
	off += 2
	putString(buf[off:], passBytes)
	return buf
}

// DecodeConnectPayload deserializes a CONNECT payload.
func DecodeConnectPayload(data []byte) (ConnectPayload, error) {
	var p ConnectPayload
	off := 0
	host, n, err := getString(data, off)
	if err != nil {
		return p, fmt.Errorf("decode connect host: %w", err)
	}
	p.Host = host
	off += n
	if off+2 > len(data) {
		return p, errors.New("decode connect: missing port")
	}
	p.Port = binary.BigEndian.Uint16(data[off:])
	off += 2
	pass, _, err := getString(data, off)
	if err != nil {
		return p, fmt.Errorf("decode connect password: %w", err)
	}
	p.Password = pass
	return p, nil
}

// EncodeErrorPayload serializes an ERROR payload.
func EncodeErrorPayload(p ErrorPayload) []byte {
	msgBytes := []byte(p.Message)
	buf := make([]byte, 2+2+len(msgBytes))
	binary.BigEndian.PutUint16(buf[0:], p.Code)
	putString(buf[2:], msgBytes)
	return buf
}

// DecodeErrorPayload deserializes an ERROR payload.
func DecodeErrorPayload(data []byte) (ErrorPayload, error) {
	var p ErrorPayload
	if len(data) < 2 {
		return p, errors.New("decode error: payload too short")
	}
	p.Code = binary.BigEndian.Uint16(data[0:2])
	msg, _, err := getString(data, 2)
	if err != nil {
		return p, fmt.Errorf("decode error message: %w", err)
	}
	p.Message = msg
	return p, nil
}

// EncodeFrame serializes a frame into bytes.
func EncodeFrame(f Frame) ([]byte, error) {
	if len(f.Payload) > MaxPayloadSize {
		return nil, ErrPayloadTooLarge
	}
	buf := make([]byte, HeaderSize+len(f.Payload))
	binary.BigEndian.PutUint32(buf[0:4], Magic)
	buf[4] = f.Version
	buf[5] = f.Type
	binary.BigEndian.PutUint16(buf[6:8], f.Flags)
	binary.BigEndian.PutUint32(buf[8:12], uint32(len(f.Payload)))
	copy(buf[HeaderSize:], f.Payload)
	return buf, nil
}

// DecodeFrame reads and decodes a single frame from r.
func DecodeFrame(r io.Reader) (Frame, error) {
	header := make([]byte, HeaderSize)
	if _, err := io.ReadFull(r, header); err != nil {
		return Frame{}, err
	}
	magic := binary.BigEndian.Uint32(header[0:4])
	if magic != Magic {
		return Frame{}, ErrInvalidMagic
	}
	ver := header[4]
	if ver != Version {
		return Frame{}, ErrInvalidVersion
	}
	length := binary.BigEndian.Uint32(header[8:12])
	if length > MaxPayloadSize {
		return Frame{}, ErrPayloadTooLarge
	}
	payload := make([]byte, length)
	if length > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return Frame{}, err
		}
	}
	return Frame{
		Version: ver,
		Type:    header[5],
		Flags:   binary.BigEndian.Uint16(header[6:8]),
		Payload: payload,
	}, nil
}

func putString(buf []byte, s []byte) int {
	binary.BigEndian.PutUint16(buf[0:2], uint16(len(s)))
	copy(buf[2:], s)
	return 2 + len(s)
}

func getString(data []byte, off int) (string, int, error) {
	if off+2 > len(data) {
		return "", 0, errors.New("missing string length")
	}
	length := int(binary.BigEndian.Uint16(data[off : off+2]))
	off += 2
	if off+length > len(data) {
		return "", 0, errors.New("string length exceeds payload")
	}
	return string(data[off : off+length]), 2 + length, nil
}

File: worker/src/protocol/frame.ts
export const MAGIC = 0x47535031; // "GSP1"
export const VERSION = 1;
export const HEADER_SIZE = 12;
export const MAX_PAYLOAD_SIZE = 16 * 1024 * 1024;

export const FrameType = {
  CONNECT: 1,
  DATA: 2,
  PING: 3,
  PONG: 4,
  CLOSE: 5,
  ERROR: 6,
} as const;

export type FrameTypeValue = (typeof FrameType)[keyof typeof FrameType];

export const ErrorCode = {
  AUTH_FAILED: 1001,
  INVALID_TARGET: 1002,
  CONNECT_FAILED: 1003,
  RATE_LIMITED: 1004,
  INVALID_FRAME: 1005,
  INTERNAL: 1006,
} as const;

export interface Frame {
  version: number;
  type: FrameTypeValue;
  flags: number;
  payload: Uint8Array;
}

export interface ConnectPayload {
  host: string;
  port: number;
  password: string;
}

export interface ErrorPayload {
  code: number;
  message: string;
}

export function encodeConnectPayload(p: ConnectPayload): Uint8Array {
  const hostBytes = new TextEncoder().encode(p.host);
  const passBytes = new TextEncoder().encode(p.password);
  const buf = new Uint8Array(2 + hostBytes.length + 2 + 2 + passBytes.length);
  const view = new DataView(buf.buffer);
  let off = 0;
  off = putString(buf, view, off, hostBytes);
  view.setUint16(off, p.port, false);
  off += 2;
  putString(buf, view, off, passBytes);
  return buf;
}

export function decodeConnectPayload(data: Uint8Array): ConnectPayload {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const [host, off1] = getString(data, view, 0);
  if (off1 + 2 > data.length) {
    throw new Error("connect payload missing port");
  }
  const port = view.getUint16(off1, false);
  const [password] = getString(data, view, off1 + 2);
  return { host, port, password };
}

export function encodeErrorPayload(p: ErrorPayload): Uint8Array {
  const msgBytes = new TextEncoder().encode(p.message);
  const buf = new Uint8Array(2 + 2 + msgBytes.length);
  const view = new DataView(buf.buffer);
  view.setUint16(0, p.code, false);
  putString(buf, view, 2, msgBytes);
  return buf;
}

export function decodeErrorPayload(data: Uint8Array): ErrorPayload {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  if (data.length < 2) {
    throw new Error("error payload too short");
  }
  const code = view.getUint16(0, false);
  const [message] = getString(data, view, 2);
  return { code, message };
}

export function encodeFrame(frame: Frame): Uint8Array {
  if (frame.payload.length > MAX_PAYLOAD_SIZE) {
    throw new Error("payload too large");
  }
  const buf = new Uint8Array(HEADER_SIZE + frame.payload.length);
  const view = new DataView(buf.buffer);
  view.setUint32(0, MAGIC, false);
  buf[4] = frame.version;
  buf[5] = frame.type;
  view.setUint16(6, frame.flags, false);
  view.setUint32(8, frame.payload.length, false);
  buf.set(frame.payload, HEADER_SIZE);
  return buf;
}

export function decodeFrame(data: Uint8Array): Frame {
  if (data.length < HEADER_SIZE) {
    throw new Error("frame too short");
  }
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const magic = view.getUint32(0, false);
  if (magic !== MAGIC) {
    throw new Error("invalid magic");
  }
  const version = data[4];
  if (version !== VERSION) {
    throw new Error("unsupported version");
  }
  const length = view.getUint32(8, false);
  if (length > MAX_PAYLOAD_SIZE) {
    throw new Error("payload too large");
  }
  if (data.length < HEADER_SIZE + length) {
    throw new Error("incomplete frame");
  }
  return {
    version,
    type: data[5] as FrameTypeValue,
    flags: view.getUint16(6, false),
    payload: data.slice(HEADER_SIZE, HEADER_SIZE + length),
  };
}

function putString(buf: Uint8Array, view: DataView, off: number, s: Uint8Array): number {
  view.setUint16(off, s.length, false);
  buf.set(s, off + 2);
  return off + 2 + s.length;
}

function getString(data: Uint8Array, view: DataView, off: number): [string, number] {
  if (off + 2 > data.length) {
    throw new Error("missing string length");
  }
  const length = view.getUint16(off, false);
  off += 2;
  if (off + length > data.length) {
    throw new Error("string exceeds payload");
  }
  const str = new TextDecoder().decode(data.slice(off, off + length));
  return [str, off + length];
}
